Add support to meshio readers and writers

keurfonluu opened the pull request:
- Added: option to read files using [``meshio``](https://github.com/nschloe/meshio) (Python library to convert unstructured grids from one format to another).

```python
import pyvista
grid = pyvista.read("test.f3grid")
grid.plot(show_edges = True)
```

![meshio2pyvista](https://user-images.githubusercontent.com/16566206/69604861-7b567080-0fd3-11ea-9947-bb811a910a3a.png)

Example FLAC3D grid file (supported by [``meshio``](https://github.com/nschloe/meshio) but not by ``pyvista``)
[test.zip](https://github.com/pyvista/pyvista/files/3890345/test.zip)

- Added: function to save files using [``meshio``](https://github.com/nschloe/meshio).

```python
import pyvista
import numpy
values = numpy.linspace(0, 10, 1000).reshape((20, 5, 10))
grid = pyvista.UniformGrid()
grid.dimensions = values.shape
grid.origin = (100, 33, 55.6)
grid.spacing = (1, 5, 2)
grid.point_arrays["values"] = values.flatten(order="F")
pyvista.save_meshio("out.msh", grid)
```

Structured grid example exported by [``meshio``](https://github.com/nschloe/meshio) and opened in Gmsh:
![pv_gmsh](https://user-images.githubusercontent.com/16566206/70202629-171c6680-16cf-11ea-8757-6fbb95ee6cbe.png)

banesullivan commented:
hi and welcome, @keurfonluu! thanks so much for getting this going as we have discussed supporting `meshio` readers before (reference https://github.com/nschloe/meshio/issues/372) but never got around to it. So far this is looking good and i have a few initial thoughts:

This PR will require some thought from the @pyvista/developers team as it introduces a new dependency. Since `meshio` is pure Python and has no dependencies beyond numpy, it would be a very lightweight addition. Also, we may want to leverage `meshio`’s writers for the `save()` method on PyVista meshes.

I’m also wondering  if we should incorporate the `meshio` readers directly in PyVista’s `read()` function to streamline this further - checking if we support a file based on the extension then checking `meshio` for support. 

@nschloe, would you have a few minutes to spare for us on this PR to double check the conversion from a `meshio.Mesh` to a `vtk.vtkUnstructuredGrid` object?

GuillaumeFavelier commented:
This is very nice to have and I think it increases the reach of `pyvista` but do you think it would be possible to integrate it while keeping it as an optional dependency for example?

nschloe commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +235,58 @@ def read_exodus(filename,
+        vtk_type = meshio_to_vtk_type[k]
+        numnodes = vtk_type_to_numnodes[vtk_type]
+        offset += [len(offset)+i*(numnodes+1) for i in range(len(v))]
+        cells += numpy.hstack((numpy.full((len(v), 1), numnodes), v)).ravel().tolist()
```
The `.tolist()` doesn't seem to be necessary. Just `numpy.concatenate` instead of `array` after the loop.

nschloe commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +235,58 @@ def read_exodus(filename,
+        if k in mesh.cell_data.keys():
+            for kk, vv in mesh.cell_data[k].items():
+                if kk in cell_data:
+                    cell_data[kk] += vv.tolist()
```
Same here.

nschloe commented:
Added some comments, looks good to me otherwise. No surprise, @keurfonluu is an excellent programmer.

GuillaumeFavelier commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +235,58 @@ def read_exodus(filename,
+def read_meshio(filename, **kwargs):
+    """Read any mesh file using meshio."""
+    # Import meshio
+    import meshio
```
I would put this in a `try` statement in case it's not installed and warn the user accordingly to make it optional.

GuillaumeFavelier commented:
I know this is not related but [greetings failed](https://github.com/pyvista/pyvista/pull/472/checks?check_run_id=320692386) :sweat_smile: with: 
```
Resource not accessible by integration
Docker run failed with exit code 1
```
Does someone remember it actually greeting once? I don't know how it works.

GuillaumeFavelier commented:
And finally could you add a test to ensure consistency and avoid regression?

akaszynski commented:
I think since meshio is a pure python module including it as a direct dependency wouldn't be an issue.  We should probably just add its functionality to `read` and `save` rather than an additional function of `read_meshio`.

Perhaps we could add `_read_meshio` internally, and if we want to make it an optional dependency, request the user install `meshio` when attempting to read the applicable files using the `read` function.

banesullivan commented:
FYI, we’d need meshio to join conda-forge if we make it a hard dependency. I can set that up if needed

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

nschloe commented:
@banesullivan  https://github.com/conda-forge/meshio-feedstock

banesullivan commented:
Ah thanks @nschloe! for some reason I didn’t see that

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

keurfonluu commented:
I modified the PR with inputs from @nschloe and @GuillaumeFavelier.

I don't know which kind of test we could do for that, but it is probably better to add it when you decided whether to integrate it directly into ``pyvista.read``. About that, guessing the file format from the extension might be problematic since many mesh formats share the same extension (e.g. ``msh`` for Gmsh and Ansys, ``inp`` for Abaqus and AVS...), that's why ``meshio.read`` has an argument ``file_format`` to specifically specify which format to use.

akaszynski commented:
I'd like to add this right to `pyvista.read` and include the `file_format` parameter so that `pyvista.read` operates similarly to `meshio.read`.  @banesullivan, what are your thoughts?

banesullivan commented:
> I'd like to add this right to pyvista.read and include the file_format parameter so that pyvista.read operates similarly to meshio.read. @banesullivan, what are your thoughts?

Agreed - I would like to have all supported file types to be able to be loaded with the `pyvista.read()` method to keep things as simple as possible for users. For implementation, I think we should still keep the `meshio` reading function separate but if a `file_format` argument is given as anything other than `None`, we call that meshio routine.

Also, it would be good to add an additional try statement here:

https://github.com/pyvista/pyvista/blob/60c96c3903deff4acde2f3f19996fcc8f5511a26/pyvista/utilities/fileio.py#L186-191

to find a mapping between the extension of the given file and the `meshio` format identifier using:

https://github.com/nschloe/meshio/blob/590b5cdd2dd761192cb44f352e706c5050d38b84/meshio/_helpers.py#L101

This would add immediate default support for a lot of formats, then for formats with duplicated extensions, the user can still pass that `file_format` parameter

@nschloe, can you make `_extension_to_filetype` publically available in `meshio`? I can't seem to access it. Or is this just name mangled?

---- 

I'd also like this to land on the next release, `0.23.0`, which is a bit overdue so let me know where I can help @keurfonluu

banesullivan commented:
My bad, @nschloe,  the `_extension_to_filetype ` is available under `meshio._helpers. _extension_to_filetype ` in the latest release (I was in an old virtual env when trying this out)

keurfonluu commented:
> For implementation, I think we should still keep the `meshio` reading function separate but if a `file_format` argument is given as anything other than `None`, we call that meshio routine.
> 
> Also, it would be good to add an additional try statement here:
> 
> https://github.com/pyvista/pyvista/blob/60c96c3903deff4acde2f3f19996fcc8f5511a26/pyvista/utilities/fileio.py#L186-L191
> 
> to find a mapping between the extension of the given file and the `meshio` format identifier using:
> 
> https://github.com/nschloe/meshio/blob/590b5cdd2dd761192cb44f352e706c5050d38b84/meshio/_helpers.py#L101
> 
> This would add immediate default support for a lot of formats, then for formats with duplicated extensions, the user can still pass that `file_format` parameter

Okay, so I updated the PR with your suggestions.

I added a ``file_format`` argument to ``pyvista.read`` that first calls ``pyvista._read_meshio`` if provided. Otherwise, ``pyvista._read_meshio`` is called as a last resort.

IMHO, keeping ``pyvista.read_meshio`` along with this new option in ``pyvista.read`` is redundant, so I made it private. I can update this PR if you still want to keep it public anyway.

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +251,57 @@ def read_exodus(filename,
     return pyvista.wrap(reader.GetOutput())
+
+
+def _read_meshio(filename, file_format = None):
```
Can you make this a public method much like the `read_exodus` method incase a user or downstream project wants to use just this routine for file reading

banesullivan replied:
i.e. no leading underscore

keurfonluu replied:
Done!

banesullivan commented:
> IMHO, keeping pyvista.read_meshio along with this new option in pyvista.read is redundant, so I made it private. I can update this PR if you still want to keep it public anyway.

I put a comment in there before seeing this. I think it would be best to keep it public just in casethere is a need for it. Otherwise, I agree - it's redundant.

> I added a file_format argument to pyvista.read that first calls pyvista._read_meshio if provided. Otherwise, pyvista._read_meshio is called as a last resort.

Perfect! Thanks!

banesullivan commented:
@keurfonluu, do you know how to convert back to `meshio` from an unstructured grid? Could we add a method in the `fileio` module called `save_meshio()` that will take any PyVista dataset and call [`.cast_to_unstructured_grid()`](https://docs.pyvista.org/core/common.html#pyvista.Common.cast_to_unstructured_grid) then save that mesh using `meshio`?

keurfonluu commented:
> @keurfonluu, do you know how to convert back to `meshio` from an unstructured grid? Could we add a method in the `fileio` module called `save_meshio()` that will take any PyVista dataset and call [`.cast_to_unstructured_grid()`](https://docs.pyvista.org/core/common.html#pyvista.Common.cast_to_unstructured_grid) then save that mesh using `meshio`?

I'll look into that. I just need to be more familiar with ``pyvista.UnstructuredGrid``.

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -187,7 +199,13 @@ def read(filename, attrs=None):
+                from meshio._helpers import _extension_to_filetype
+                return read_meshio(filename, _extension_to_filetype[ext])
+            except:
+                pass
```
Perhaps we should throw an error or warning message  here if `meshio` is not installed - if it is not installed, then a user will just see our uninformative error at the end of this method

keurfonluu replied:
Will ``meshio`` be a dependency of ``pyvista``? If so, it should be installed when installing ``pyvista``, right?

banesullivan replied:
Ah yes - can you go ahead and add `meshio` to the `setup.py` and in the `requirements.txt`? Do you know if we should specify above a specific version?

banesullivan replied:
And can you also list it in the "getting started" page in the docs?

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -153,13 +156,22 @@ def read(filename, attrs=None):
+        try:
+            return read_meshio(filename, file_format)
+        except:
+            pass
```
Would `meshio` throw an informative error worth showing the user? if so, this bare except will skip over it

Nicztin replied:
> +        try:
+            return read_meshio(filename, file_format)
+        except:
+            pass
Would meshio throw an informative error worth showing the user? if so, this bare except will skip over it

keurfonluu replied:
Yes, ``meshio`` is throwing some ``AssertionError`` depending on the error. I can remove this ``try/except`` if you think it would be better to show the actual error.

banesullivan replied:
Let's take away the try/except - that way if the user passes a non-working `file_format` argument they will be alerted by `meshio`

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -187,7 +199,13 @@ def read(filename, attrs=None):
-            pass
+            # Attempt read with meshio
+            try:
+                from meshio._helpers import _extension_to_filetype
```
Can we put this in the `read_meshio` method and default the `file_format` argument to `None`, then if it is `None`, look it up here?

banesullivan replied:
Moving this would resolve my comment below about the warning message if meshio is not installed

keurfonluu replied:
Might be useless. I'll remove it since this dict is used to guess type in ``meshio``.

Nicztin commented:
What is this?

Send from N.E.22485Ct, CA92587 


From: Bane Sullivan
Sent: Wednesday, December 4, 2019 8:19 AM
To: pyvista/pyvista
Cc: Subscribed
Subject: Re: [pyvista/pyvista] Add support to meshio readers (#472)

@banesullivan commented on this pull request.

In pyvista/utilities/fileio.py:
> @@ -233,3 +251,57 @@ def read_exodus(filename,
 
     reader.Update()
     return pyvista.wrap(reader.GetOutput())
+
+
+def _read_meshio(filename, file_format = None):
i.e. no leading underscore
—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +251,57 @@ def read_exodus(filename,
+    )
+    
+    # Read mesh file
+    mesh = meshio.read(filename, file_format)
```
Wait, if file_format is `None` from the default argument here, will `meshio` automatically chose a reader based on the `_extension_to_filetype` dict? If so, we don't need to use that dict anywhere on PyVista's side

keurfonluu replied:
Yes, that's right. ``meshio`` also has its own format guessing feature if not provided. Probably better to not use it in ``pyvista.read``.

banesullivan replied:
Okay, let's completely rely on `meshio`'s format guessing technique and remove PyVista's use of the `_extension_to_filetype`

banesullivan commented:
https://github.com/pyvista/pyvista/pull/472#issuecomment-561436644:
> What is this?

Hi @Nicztin, are you asking why you received that email? You are watching the `pyvista` repository which likely notifies you of all conversations

keurfonluu commented:
I added the function to save meshes with ``meshio``. Also fixed a couple of bugs in ``pyvista.read_meshio``.

There is one "ugly" thing I had to do in the writer in order to convert voxels to hexahedra since ``meshio`` does not handle voxels. I don't know why the conversion is not performed in ``meshio``, but there might be a good reason, right @nschloe ?

nschloe commented:
>  but there might be a good reason, right?

Well, we don't transform pixels into quads either. I guess we _could_ do that, but it seems more like a gimmick than a serious application.

banesullivan commented:
This is looking great, @keurfonluu! All we need now are comprehensive tests and this will be ready to go

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +249,106 @@ def read_exodus(filename,
+    from meshio._vtk import vtk_to_meshio_type
+
+    # Cast to pyvista.UnstructuredGrid
+    mesh = mesh.cast_to_unstructured_grid()
```
```suggestion
    if not isinstance(mesh, pyvista.UnstructuredGrid):
        mesh = mesh.cast_to_unstructured_grid()
```

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +249,106 @@ def read_exodus(filename,
+
+
+def save_meshio(filename, mesh, file_format = None, **kwargs):
+    """Save mesh to file using meshio."""
```
```suggestion
    """Save mesh to file using meshio.

    Parameters
    ----------
    mesh : pyvista.Common
        Any PyVista mesh/spatial data type.
    file_format : str
        File type for meshio to save.

    """

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +249,106 @@ def read_exodus(filename,
+def save_meshio(filename, mesh, file_format = None, **kwargs):
+    """Save mesh to file using meshio."""
+    from meshio._vtk import vtk_to_meshio_type
+
```
Make sure relative paths like `"~/Desktop.foo.vti"` will work
```suggestion
    filename = os.path.abspath(os.path.expanduser(filename))
```

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -233,3 +249,106 @@ def read_exodus(filename,
+        meshio_to_vtk_type,
+        vtk_type_to_numnodes,
+    )
+    
```
Make sure relative paths like `"~/Desktop.foo.vti"` will work
```suggestion

    filename = os.path.abspath(os.path.expanduser(filename))
```

banesullivan commented on pyvista/utilities/fileio.py:
```diff
@@ -8,6 +8,10 @@
 import imageio
 
+import numpy
+
```
```suggestion
import os
```

keurfonluu commented:
> This is looking great, @keurfonluu! All we need now are comprehensive tests and this will be ready to go

I added a simple test that saves the beam example file and read it back using ``meshio``, and check that ``points``, ``cells``, ``point_arrays`` and ``cell_arrays`` are still the same.

Do you have any idea of other tests I can add?

banesullivan commented:
@keurfonluu, that sounds great! Let's also save the plane example (`examples.load_airplane()`) and the uniform grid (`examples.load_uniform()`) just to make sure other mesh types are working. We don't need to test all the functionality of `meshio` but just that the code here is working to send meshes back and forth with `meshio`.

We may also want to test that common errors/issue are handled like if an undefined format is passed the reader/writer.

banesullivan commented on tests/test_meshio.py:
```diff
@@ -0,0 +1,27 @@
+    # Save and read reference mesh using meshio
+    pyvista.save_meshio("test_meshio.vtk", mesh_ref)
+    mesh = pyvista.read_meshio("test_meshio.vtk")
+    os.remove("test_meshio.vtk")
```
Can you use `tmpdir` with PyTest much like other tests: https://github.com/pyvista/pyvista/blob/60c96c3903deff4acde2f3f19996fcc8f5511a26/tests/test_polydata.py#L256

keurfonluu commented:
> @keurfonluu, that sounds great! Let's also save the plane example (`examples.load_airplane()`) and the uniform grid (`examples.load_uniform()`) just to make sure other mesh types are working. We don't need to test all the functionality of `meshio` but just that the code here is working to send meshes back and forth with `meshio`.
> 
> We may also want to test that common errors/issue are handled like if an undefined format is passed the reader/writer.

All done.

For the uniform example, one part of the test converts the voxels to hexahedra (as in ``pyvista.save_meshio``) to allow the comparison, which means that the conversion is not conservative for uniform grids with voxels (and pixels).

agramfort commented on pyvista/utilities/fileio.py:
```diff
@@ -8,6 +8,10 @@
 
+import numpy
+
+import meshio
```
I am a bit puzzled you add this as a hard dependency. I would go with nested imports.
Also I would recommend to have a travis build that runs with only hard dependencies (numpy + vtk).

my 2c

akaszynski replied:
It's a pure python module, so there's not much harm done by adding `meshio` as a dependency, but it's something to consider as I too don't like having a ton of dependencies.  However, I think it's worth the small additional install time for the user to be able to seamlessly use `meshio` within `pyvista`.

banesullivan replied:
I'm in agreement with @akaszynski here - `meshio` is lightweight enough and the benefits of the supported file types it will add make it worth-while. 

More supported file types ▶️  more interest ▶️ more users ▶️ more contributors ▶️ better/more stable PyVista

agramfort replied:
I am not saying you should not support it I am suggesting you make it a nested import. In MNE for example we won't need this. Any dependency will have a cost for you in the future. For example you need to do a debian package, then you'll have to package 2 libraries separately. If meshio looses is core dev you'll have to maintain it or find an alternative etc. I don't have the authority here to make this call, but really think twice before adding deps. FYI mne and scikit-learn have as minimal deps numpy and scipy. PyVista is really for me a thin layer of pure python code on top of vtk so vtk+numpy should be all I need. my 2c

nschloe replied:
> For example you need to do a debian package, then you'll have to package 2 libraries separately.

https://tracker.debian.org/pkg/python-meshio

agramfort replied:
yes I saw it's packaged. again feel free to ignore my comment if you consider it inappropriate.

akaszynski replied:
That's cool that it's a debian package!

In many ways, I'm with you @agramfort.  I also don't have plans (at the moment) for using `meshio`, but if it can read in some ANSYS files, that might be something I look at in the future.  A nested import with an optional dependency would work just fine for me.  However, considering how huge `numpy` and `vtk` are, I don't think we should worry about a pure python package.  Correct me if I'm wrong, but I don't see where a debian package would be used for a pure python install if you're installing it using `pip`, and packaging it with conda should also take care of the dependency issues.

I agree that adding another dependency complicates things, and this is exactly why the dependencies for writing movies were left as an optional dependency with a nested import.  Installing it for me requires a *.deb package, and Windows requires extra work as well, and it's already tough to install VTK for some.

Since we should be mostly concerned with balancing the install/use experience, I see no added cost including a lightweight pure python dependency as this saves the user from going back through and running `pip install meshio` after they get a `ModuleNotFoundError`.  We already install `scooby` and `appdirs` which are optional for most plotting functions, but the extra step to install them when the user needs to print out a system report or download an example file wouldn't seem polished for me.

GuillaumeFavelier replied:
In general, I'm in favour of [*"Always implement things when you actually need them, never when you just foresee that you need them."*](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

akaszynski commented:
Feel free to merge @keurfonluu.  Thanks for adding this!

banesullivan commented:
> Feel free to merge @keurfonluu. Thanks for adding this!

I don't think @keurfonluu has merge access - when you're read, @keurfonluu, let me know and I will merge

keurfonluu commented:
> > Feel free to merge @keurfonluu. Thanks for adding this!
> 
> I don't think @keurfonluu has merge access - when you're ready, @keurfonluu, let me know and I will merge

I have nothing else to add. Feel free to merge this PR. Thanks!
